include/qinterface.hpp

```diff
@@ -110,9 +114,24 @@ class QInterface {
     virtual real1 Rand() { return rand_distribution(*rand_generator); }
     virtual void SetRandomSeed(uint32_t seed) { rand_generator->seed(seed); }
 
+    virtual void Apply2x2(bitCapInt offset1, bitCapInt offset2, const complex* mtrx, const bitLenInt bitCount,
+        const bitCapInt* qPowersSorted, bool doCalcNorm)
+    {
+        throw "Apply2x2 not implemented in interface";
```

Comment:
Would a pure virtual via `= 0;` at the end work?

Reply:
Stylistically and in terms of function, I think this is preferable. Protected methods are used internally by QInterface, but QEngineOCLMulti doesn't necessarily implement them, for example. The bodies with exceptions thrown keep QEngineOCLMulti and other potential engines from having to override pure virtual protected methods they don't actually need. Meanwhile, if one specifically intends to avoid using a particular method, the exception thrown lets us know if we've accidentally used it in an engine without overriding it with a real implementation.

The pattern that naturally arose seemed to be, engine-specific public methods should be pure virtual, since they're all required, but protected methods relied on internally by QInterface are potentially optional and should throw exceptions for not being implemented if they rely on a specific manipulation of the state vector, so that needless method bodies don't need to added to engines like QEngineOCLMulti.